reactor3/src/test/java/com/linecorp/armeria/common/reactor3/RequestContextPropagationFluxTest.java

```diff
@@ -0,0 +1,473 @@
+                    .expectSubscriptionMatches(s -> ctxExists(ctx))
+                    .expectNextMatches(s -> ctxExists(ctx) && "baz".equals(s))
+                    .verifyComplete();
+    }
+
+    @Test
+    void ctxShouldBeCleanUpEvenIfErrorOccursDuringReactorOperationOnSchedulerThread() {
+        // Given
+        final ClientRequestContext ctx = newContext();
+        final Flux<String> flux;
+
+        // When
+        flux = Flux.just("Hello", "Hi")
+                   .subscribeOn(Schedulers.single())
+                   .delayElements(Duration.ofMillis(1000))
+                   .map(s -> {
+                       if (s.equals("Hello")) {
+                           throw new RuntimeException();
+                       }
+                       return s;
+                   })
+                   .contextWrite(Context.of(RequestContextAccessor.accessorKey(), ctx));
+
+        // Then
+        StepVerifier.create(flux)
+                    .expectError(RuntimeException.class)
+                    .verify();
+
+        final Flux<String> toVerifyFlux = Flux.just("Dummy")
+                                        .subscribeOn(Schedulers.single())
+                                        .doOnNext(s -> assertThat(ctxExists(ctx)).isFalse());
+
+        StepVerifier.create(toVerifyFlux)
+                    .expectNext("Dummy")
+                    .verifyComplete();
+    }
+
+    @Test
+    void ctxShouldBeCleanUpEvenIfErrorOccursDuringReactorOperationOnMainThread() {
+        // Given
+        final ClientRequestContext ctx = newContext();
+        final Flux<String> flux;
+
+        // When
+        flux = Flux.just("Hello", "Hi")
+                   .map(s -> {
+                       if (s.equals("Hello")) {
+                           throw new RuntimeException();
+                       }
+                       return s;
+                   })
+                   .contextWrite(Context.of(RequestContextAccessor.accessorKey(), ctx));
+
+        // Then
+        StepVerifier.create(flux)
+                    .expectError(RuntimeException.class)
+                    .verify();
+
+        assertThat(ctxExists(ctx)).isFalse();
+    }
+
+    private static <T> Flux<T> addCallbacks(Flux<T> flux, ClientRequestContext ctx) {
+        return flux.doFirst(() -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnSubscribe(s -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnRequest(l -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnNext(foo -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnComplete(() -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnEach(s -> assertThat(ctxExists(ctx)).isTrue())
+                   .doOnError(t -> assertThat(ctxExists(ctx)).isTrue())
+                   .doAfterTerminate(() -> assertThat(ctxExists(ctx)).isTrue())
+                   .contextWrite(Context.of(RequestContextAccessor.accessorKey(), ctx));
```

Comment:
Could you also add tests that use `contextCapture`?

Reply:
Of course, Sure!  I didn't realized about `contextCapture` at all. 🙇‍♂️ 
I added test cases that use `contextCapture()` as well. 
The test cases are provided by `@ParameterizedTest` and `@MethodSource("...")`

- Modify `addCallbacks()` to get `shouldContextWrite` and `shouldContextCapture`. 
- If `shouldContextCapture` is true, tests are executed with `try (SafeCloseable ignored = ctx.push())`. Otherwise, tests are executed without `try (SafeCloseable ignored = ctx.push())`.